### PR TITLE
Add sender companion account and their balances to account's data model

### PR DESCRIPTION
**Pull request by hbulgarini**
First step for the work described in #224.

This PR is just creating the mechanism to add companion account derived process and the balance subscriptions to the data model only. In order to keep PR small , this PR is not including the actual use of this new states.

@goldsteinsveta  this change might simplify a little bit the logic for `Sender` component.
However i still don't have clear how to approach subscriptions for accounts listed in the dropdown. Being disscussed.

### CODE REVIEW

**Review by wirednkod — APPROVED** [review]
lgtm

File: src/hooks/accounts/useSourceAccountsBalances.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    ss58Format: configs.ss58Format,
+    address: accountState.account?.address || '',
+    bridgeId: getBridgeId(configs, chainName)
+  };
+  const companionAccount = getDeriveAccount(toDerive);
+  const companionBalance = useBalance(targetApi, companionAccount || '', true);
+
+  useEffect(() => {
+    dispatchAccount(AccountActionCreators.setSenderBalances(accountBalance, companionBalance));
+    dispatchAccount(AccountActionCreators.setSenderCompanionAccount(companionAccount));
```
Comment (tomusdrw) [inline src/hooks/accounts/useSourceAccountsBalances.ts:49]: This is always a smell to me. If things are always changed together it should be one reducer action.

Comment (tomusdrw) [reply src/hooks/accounts/useSourceAccountsBalances.ts:49]: Alternatively maybe split this into two separate `useEffects`?

File: src/hooks/accounts/useSourceAccountsBalances.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  } = useSourceTarget();
+
+  const accountBalance = useBalance(sourceApi, accountState.account?.address || '', true);
+
+  const toDerive = {
+    ss58Format: configs.ss58Format,
+    address: accountState.account?.address || '',
+    bridgeId: getBridgeId(configs, chainName)
+  };
+  const companionAccount = getDeriveAccount(toDerive);
```
Comment (tomusdrw) [inline src/hooks/accounts/useSourceAccountsBalances.ts:44]: IMHO this should be part of the reducer logic and not the hook.

File: src/contexts/AccountContextProvider.tsx [hunk 3]
```diff
@@ -46,19 +46,24 @@ export function useUpdateAccountContext() {
...
+  const [accountState, dispatchAccount] = useReducer(accountReducer, {
     account: null,
-    accounts: []
+    accounts: [],
+    companionAccount: null,
+    senderAccountBalance: null,
+    senderCompanionAccountBalance: null
   });
 
+  useSourceAccountsBalances(accountState, dispatchAccount);
```
Comment (tomusdrw) [inline src/contexts/AccountContextProvider.tsx:57]: This is cool, but the only issue I have now is more about directory structure. I imagine that a generic `hooks` folder contains hooks that can be used in any place without any additional dependencies. Now, this particular hook can only be used in conjunction with a reducer and pretty much only in this one place - however by looking into `hooks` directory you won't be able to distinguish "regular hooks" and "helper/specialised hooks" - this long term will imho lead to confusion and "the code is complicated" feeling.

To overcome that we can either put such hooks in a different place, or have the hook `useReducer` internally. The latter is maybe a bit weird, cause if we happen to actually use that hook in other places than the context you will have a separate copy of the state, but well that might actually be a feature really.

File: src/util/getDeriveAccount.ts [hunk 4]
```diff
@@ -26,6 +26,9 @@ interface Data {
 }
 
 export default function getDeriveAccount({ ss58Format = 42, bridgeId, address }: Data): string {
+  if (!address) {
+    return address;
+  }
```
Comment (tomusdrw) [inline src/util/getDeriveAccount.ts:31]: This is a bit weird, why special case empty string?

**Review by tomusdrw — APPROVED** [review]
lgtm! Some grumbles in the comments, but it's just old man's whining :)

File: src/hooks/accounts/useSourceAccountsBalances.ts (hunk 1 above)
Comment (hbulgarini) [reply src/hooks/accounts/useSourceAccountsBalances.ts:49]: no, this should be moved to reducer as you suggested.

File: src/hooks/accounts/useSourceAccountsBalances.ts (hunk 2 above)
Comment (hbulgarini) [reply src/hooks/accounts/useSourceAccountsBalances.ts:44]: totally, i think this came from moving the hooks.

File: src/contexts/AccountContextProvider.tsx (hunk 3 above)
Comment (hbulgarini) [reply src/contexts/AccountContextProvider.tsx:57]: it's a valid point which makes sense to me. I'm wondering if below the `hooks` directory, i could create a `context` sub folder and place this kind of hooks. Additionally i think it is interesting moving the `useReducer` to this specific dedicated hook.

File: src/util/getDeriveAccount.ts (hunk 4 above)
Comment (hbulgarini) [reply src/util/getDeriveAccount.ts:31]: sorry , i'm not getting your question. Do you mean why i'm handling the empty address case?

**Comment by hbulgarini** [comment]
> lgtm! Some grumbles in the comments, but it's just old man's whining :)

hell no!  it is great feedback that i will address right now thanks!

**Pull request merged**